src/lib/plugin/utils.ts

```diff
@@ -16,32 +19,89 @@ const log = rootLogger.child('plugin')
-  try {
-    const packageJson = require(plugin.packageJsonPath) as PackageJson
-
-    if (!packageJson.name) {
-      fatal(`One of your plugin has a missing required \`name\` property in its package.json`, {
-        packageJsonPath: plugin.packageJsonPath,
-        packageJson,
-      })
-    }
+export async function getPluginManifest(plugin: Plugin): Promise<Manifest> {
+  const errPackageJson = tryCatch(() => require(plugin.packageJsonPath) as PackageJson, toError)
 
-    return {
-      ...plugin,
-      name: packageJson.name,
-      packageJson,
-    }
-  } catch (error) {
+  if (isLeft(errPackageJson)) {
     fatal(
-      stripIndent`
-       An error occured when reading the package.json of one of your Nexus plugin:
-
-       ${error.stack ?? error}
-    `,
+      createGetManifestError(
+        addErrorMessageContext(`Failed to read the the plugin's package.json file.`, errPackageJson.left)
+      ),
       { plugin }
     )
   }
+
+  const packageJson = errPackageJson.right
+
+  if (!packageJson.name) {
+    fatal(createGetManifestError(new Error(`\`name\` property is missing in package.json`)), {
+      packageJson: {
+        data: packageJson,
+        path: plugin.packageJsonPath,
+      },
+    })
+  }
+
+  if (!packageJson.main) {
+    fatal(createGetManifestError(new Error(`\`main\` property is missing in package.json`)), {
+      packageJson: {
+        data: packageJson,
+        path: plugin.packageJsonPath,
+      },
+    })
+  }
+
+  const validatedPackageJson = packageJson as ValidatedPackageJson
+
+  const [worktime, runtime, testtime] = await Promise.all([
+    checkForConventionalDimensionEntrypoint('worktime', validatedPackageJson, plugin),
+    checkForConventionalDimensionEntrypoint('runtime', validatedPackageJson, plugin),
+    checkForConventionalDimensionEntrypoint('testtime', validatedPackageJson, plugin),
+  ])
+
+  return {
+    name: packageJson.name,
+    settings: (plugin as any).settings ?? null,
+    packageJsonPath: plugin.packageJsonPath,
+    packageJson: validatedPackageJson,
+    worktime,
+    testtime,
+    runtime,
+  }
+}
+
+/**
+ * Get the dimension entrypoint location. Take it from the manifest input if
+ * present. Otherwise check on disk for the conventional module.
+ *
+ * The conventional dimension module location is:
+ *
+ * <project-root>/<main dir>/{runtime,worktime,testtime}.js
+ * <project-root>/<main dir>/{runtime,worktime,testtime}/index.js
+ *
+ * The location path extension is not specified to afford plugin author the
+ * index dir style. This means the location path must be passed into a
+ * node-module-resolution functino e.g. `require`. Do not pass it to a raw FS
+ * file read function.
+ */
+async function checkForConventionalDimensionEntrypoint(
+  dimensionkind: Dimension,
+  packageJson: ValidatedPackageJson,
+  plugin: Plugin
+): Promise<DimensionEntrypointLocation | null> {
+  if (plugin[dimensionkind]) return plugin[dimensionkind]!
```

Comment:
Arguably this check should be done a higher level up, and instead of passing `plugin` we should pass `packageJsonPath` here. It would narrow the inputs which is good, but, make the helper func less... helpful.